controllers/complianceeventsapi/server.go

```diff
@@ -0,0 +1,334 @@
+		}
+
+		policyFK, err := getPolicyForeignKey(r.Context(), goquDB, reqEvent.Policy)
+		if err != nil {
+			log.Error(err, "error getting policy foreign key")
+			http.Error(w, "Internal Error", http.StatusInternalServerError)
+
+			return
+		}
+
+		reqEvent.Event.PolicyID = policyFK
+
+		insert := goquDB.Insert("compliance_events").Rows(reqEvent.Event).Executor()
+
+		_, err = insert.Exec()
+		if err != nil {
+			log.Error(err, "error inserting compliance event")
+			http.Error(w, "Internal Error", http.StatusInternalServerError)
+		}
+	}
+}
+
+func getClusterForeignKey(ctx context.Context, goquDB *goqu.Database, cluster Cluster) (int, error) {
+	// Check cache
+	key, ok := clusterKeyCache.Load(cluster.ClusterID)
+	if ok {
+		return key.(int), nil
+	}
+
+	foundCluster := new(Cluster)
+
+	found, err := goquDB.From("clusters").
+		Where(goqu.Ex{"cluster_id": cluster.ClusterID}).
+		ScanStructContext(ctx, foundCluster)
+	if err != nil {
+		return 0, err
+	}
+
+	// If the row already exists
+	if found {
+		clusterKeyCache.Store(cluster.ClusterID, foundCluster.KeyID)
+
+		return foundCluster.KeyID, nil
+	}
+
+	// Otherwise, create a new row in the table
+	insert := goquDB.Insert("clusters").Returning("id").Rows(cluster).Executor()
+
+	id := new(int)
+	if _, err := insert.ScanValContext(ctx, id); err != nil {
+		return 0, err
+	}
+
+	clusterKeyCache.Store(cluster.ClusterID, *id)
+
+	return *id, nil
+}
+
+func getParentPolicyForeignKey(ctx context.Context, goquDB *goqu.Database, parent ParentPolicy) (int, error) {
+	// Check cache
+	parKey := parent.key()
+
+	key, ok := parentPolicyKeyCache.Load(parKey)
+	if ok {
+		return key.(int), nil
+	}
+
+	foundParent := new(ParentPolicy)
+
+	uniqueFieldsMatch := goqu.Ex{
+		"name": parent.Name,
+	}
+
+	if len(parent.Categories) > 0 {
+		uniqueFieldsMatch["categories"] = goqu.L("?::text[]", parent.Categories)
+	}
+
+	if len(parent.Controls) > 0 {
+		uniqueFieldsMatch["controls"] = goqu.L("?::text[]", parent.Controls)
+	}
+
+	if len(parent.Standards) > 0 {
+		uniqueFieldsMatch["standards"] = goqu.L("?::text[]", parent.Standards)
+	}
+
+	qu := goquDB.From("parent_policies").Where(uniqueFieldsMatch)
+
+	found, err := qu.ScanStructContext(ctx, foundParent)
+	if err != nil {
+		return 0, err
+	}
+
+	// If the row already exists
+	if found {
+		parentPolicyKeyCache.Store(parKey, foundParent.KeyID)
+
+		return foundParent.KeyID, nil
+	}
+
+	// Otherwise, create a new row in the table
+	insert := goquDB.Insert("parent_policies").Returning("id").Rows(parent).Executor()
+
+	id := new(int)
+	if _, err := insert.ScanValContext(ctx, id); err != nil {
+		return 0, err
+	}
+
+	parentPolicyKeyCache.Store(parKey, *id)
+
+	return *id, nil
+}
+
+func getPolicyForeignKey(ctx context.Context, goquDB *goqu.Database, pol Policy) (int, error) {
+	// Fill in missing fields that can be inferred from other fields
+	if pol.SpecHash == nil {
+		var buf bytes.Buffer
+		if err := json.Compact(&buf, []byte(*pol.Spec)); err != nil {
```

Comment:
Is this deterministic? For example, are struct fields always marshaled in the same order?

Reply:
I guess it'd be up to the API caller to make this deterministic.

Reply:
I'll check on this.

Reply:
It doesn't unmarshal and then marshal again, so the order in the request does matter. Switching the order of the keys changes the hash later.

I think we should somehow discourage users from trying to do the hash themselves: if they don't know what it is, they should just submit the whole spec, and the response should have the hash in case they want to submit more events for the same policy. But I'm not sure the ideal design to accomplish that.

Reply:
I agree to your last comment. I think we need an algorithm that orders the JSON consistently so the hash is always the same from the same JSON.

I don't really see a way around this unless we just use the `policy-templates` entry as the "official" spec and the template sync sets the spec hash as an annotation or something on the policy template (e.g. ConfigurationPolicy).

Reply:
I think this is solved with the idea of setting the `policy` ID and `parent_policy` ID as annotations on the `ConfigurationPolicy`.